### DISCUSSION TITLE
name 'ls' isn't defined

### DISCUSSION THREAD

**Discussion by stas-sl** [Carglglz/upydev#32] [Q&A] [answered]
Hi!

I'm getting the following error when trying to execute `ls` command:

```
esp8266@d1:~ $ ls
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'ls' isn't defined
```
I already run `enable_sh` and it copied upysh.py to /lib directory, do I need something else to enable `ls` command?

I'm using latest upydev version 0.4.3

**Comment by stas-sl** [comment]
Hmm. It was quite confusing, but seems like I found the issue. I used OTA micropython version which doesn't support f-strings (https://github.com/micropython/micropython/issues/8450). However I didn't see any error message when trying to `import upysh` - it just silently didn't work. The only way I figured it out was using `rshell` - it showed line number where error was.

Though, I solved this issue and most shell commands work, `ls` still giving me another error:
```
esp8266@d1:~ $ ls
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unexpected keyword argument 'gts'
```

But it works when running from REPL:
```
esp8266@d1:~ $
>>> ls
     381 boot.py
   <dir> lib
      14 webrepl_cfg.py
```

**Comment by Carglglz** [comment, marked as answer]
@stas-sl 
Hi, TLDR: add these lines of code in `boot.py` or `main.py`
```python
import sys
sys.path[:] = ['', '/lib', '/', '.frozen']
```
And that should fix it. 

This happens because MicroPython recently started to include its version of `upysh.py` as a frozen module i.e `.frozen` , you can see this with 

```
esp8266@esp8266:~ $ modules
__main__          math              ucollections      ure
_boot             micropython       ucryptolib        urequests
_onewire          neopixel          uctypes           urllib/urequest
_uasyncio         network           uerrno            uselect
_webrepl          ntptime           uhashlib          usocket
apa102            onewire           uheapq            ussl
btree             port_diag         uio               ustruct
builtins          ssd1306           ujson             usys
dht               uarray            umachine          utime
ds18x20           uasyncio/__init__ umqtt/robust      uwebsocket
esp               uasyncio/core     umqtt/simple      uzlib
flashbdev         uasyncio/event    uos               webrepl
framebuf          uasyncio/funcs    upip              webrepl_setup
gc                uasyncio/lock     upip_utarfile
inisetup          uasyncio/stream   upysh
lwip              ubinascii         urandom
Plus any modules on the filesystem
```

And default `sys.path` to look for modules is 
```
>>> import sys
>>> sys.path
['', '.frozen', '/lib', '/']

```
So calling `ls` with this `sys.path` will import the frozen module, instead of the one in `/lib` you just copied.
Then as I said above add those lines of code and it should work 👍🏼

**Reply by stas-sl** [reply]
Thanks for fast response, @Carglglz! I didn't know there is upysh in micropython itself, but your trick helped. Maybe it worth renaming your `upysh` to something else in future versions to avoid conflicts?

**Reply by Carglglz** [reply]
@stas-sl 
> Maybe it worth renaming your upysh to something else in future versions to avoid conflicts?

Yes, I will look into this for the next release, that or maybe copy `upysh.py` to `''` instead of `/lib` , that should work too... 🤔

**Comment by thermike2** [comment]
Hi  @Carglglz,
I am trying to use the PN532 NFC/RFID lib to read a card should i have an interrupt and a card is present instead of polling on an endless loop. I have connected the IRQ of the SPI module to GPIO20 on a pico but I can't make it to work. Here I have a more detailed description https://raspberrypi.stackexchange.com/questions/143344/raspberry-pi-pico-with-pn532-reading-cards-with-the-use-of-interrupt.
Can you help me a little ? What am i doing wrong? Do you have an example?
